commons-numbers-complex-arrays/src/test/java/org/apache/commons/numbers/complex/arrays/ComplexListTest.java

```diff
@@ -223,14 +222,87 @@ void testReplaceAllComplexScalarFunction() {
+            list.setReal(i, value);
+            list.setImaginary(i, value);
+            Assertions.assertEquals(value, list.get(i).getReal());
+            Assertions.assertEquals(value, list.get(i).getImaginary());
+        }
+    }
+
+    @Test
+    void testGetAndSetRealAndImaginaryIndexOutOfBoundsException() {
+        ComplexList list = new ComplexList();
+        // Empty list throws
+        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> list.getReal(0));
+        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> list.getImaginary(0));
+
+        int size = 5;
+        IntStream.range(0, size).mapToObj(i -> Complex.ofCartesian(i, -i)).forEach(list::add);
+
+        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> list.getReal(-1));
+        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> list.getReal(size));
+        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> list.getReal(size + 1));
+
+        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> list.getImaginary(-1));
+        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> list.getImaginary(size));
+        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> list.getImaginary(size + 1));
+
+        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> list.setReal(-2, 200));
+        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> list.setReal(size, 200));
+        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> list.setReal(size + 1, 200));
+
+        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> list.setImaginary(-2, 200));
+        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> list.setImaginary(size, 200));
+        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> list.setImaginary(size + 1, 200));
+    }
+
+    @Test
+    void testToArrayRealAndImaginary() {
+        ComplexList list = generateList(10);
+        double[] expectedReal = list.stream().mapToDouble(Complex::getReal).toArray();
+        double[] actualReal = list.toArrayReal();
+        Assertions.assertArrayEquals(expectedReal, actualReal);
+        double[] expectedImaginary =  list.stream().mapToDouble(Complex::getImaginary).toArray();
+        double[] actualImaginary = list.toArrayImaginary();
+        Assertions.assertArrayEquals(expectedImaginary, actualImaginary);
+    }
+
     /**
-     * Generates a list of random complex numbers of the given size.
+     * Generates a ComplexList of random complex numbers of the given size.
      * @param size number of complex numbers in the list.
-     * @return the list of random complex numbers.
+     * @return the ComplexList of random complex numbers.
      */
-    private static List<Complex> generateList(int size) {
-        return ThreadLocalRandom.current().doubles(size, -Math.PI, Math.PI)
+    private static ComplexList generateList(int size) {
```

Comment:
There are use cases for `generateList` and `generateComplexList`. The generateList would function as before and return a List implementation from the JDK. This can be assumed to work correctly. The generateComplexList would call generateList and then addAll to a new ComplexList instance. This should be asserted to be equal to the original List. This method can be used for convenience to populate a ComplexList for testing the accessor methods (get/set).

One case where you may wish to have a List that is not a ComplexList is in testing addAll. You also require a non ComplexList for any list compared for equality with ComplexList: you cannot have a list equals comparison to itself as there may be the same error in both implementations being compared.